Add a debugging section to the documentation (#1337)
trixi-framework/Trixi.jl: docs/src/development.md

mode [julia-emacs](https://github.com/JuliaEditorSupport/julia-emacs).

## Debugging
Julia offers several options for debugging. A classical debugger is available with the 
[Debugger.jl](https://github.com/JuliaDebug/Debugger.jl) package or in the 
[Julia extension for VS Code](https://www.julia-vscode.org/docs/stable/userguide/debugging/). 

> However, it can be quite slow and does currently not work properly for Trixi. The 

Author: Done in e933579e5f9a007f4e406f422a5d4f0933b62e2c

[Infiltrator.jl](https://github.com/JuliaDebug/Infiltrator.jl) package on the other hand does not 
offer all features of a full debugger, but is a fast and simple tool that allows users to set 
breakpoints to open a local REPL session and access the call stack and variables.

### Infiltrator

> The Infiltrator package provides fast breakpoints using the ```@infiltrate``` command to drop the 
> user into a local REPL session, where it is possible to access local variables, see the call stack 
> and execute statements.

Reviewer: ```suggestion
The Infiltrator package provides interactive breakpoints using the ```@infiltrate``` command, which drops the 
user into a local REPL session. From there, it is possible to access local variables, see the call stack 
and execute statements.
```

Author: Done in e933579e5f9a007f4e406f422a5d4f0933b62e2c

The package can be installed in the Julia REPL by executing 
```julia-repl
(@v1.8) pkg> add Infiltrator
```

To load the package in the Julia REPL execute
```julia-repl
julia> using Infiltrator
```

> Breakpoints can be set by adding a line with the ```@infiltrate``` macro at the respective position 
> in the code. Use [Revise](@ref interactive-use-of-julia) if you want to set and delete breakpoints 
> interactively. 

Reviewer: ```suggestion
Breakpoints can be set by adding a line with the ```@infiltrate``` macro at the respective position 
in the code. Use [Revise](@ref interactive-use-of-julia) if you want to set and delete breakpoints 
in your package code without having to restart Julia. 
```

Author: Done in e933579e5f9a007f4e406f422a5d4f0933b62e2c

!!! note 
    In a package environment the ```@infiltrate``` macro only works when used in a specific way, 
    which is why the functional form

>     ```julia
>     if isdefined(Main, :Infiltrator)
>     Main.infiltrate(@__MODULE__, Base.@locals, @__FILE__, @__LINE__)
>     end
>     ```

Reviewer 2: ```suggestion
    ```julia
    if isdefined(Main, :Infiltrator)
      Main.infiltrate(@__MODULE__, Base.@locals, @__FILE__, @__LINE__)
    end
    ```
```

Author: I am not sure what you mean? What is the benefit of using `Main.Infiltrator.infiltrate` instead of `Main.infiltrate`?

Reviewer 2: You just check whether `Main.Infiltrator` is defined. That is also the case when I execute `import Infiltrator`. But in the code, you assume that `Main.infiltrate` is defined, which is not the case after `import Infiltrator` or `using Infiltrator: Infiltrator`. You need `using Infiltrator` to get the name `Main.infiltrate`.

Author: I see, thanks for the explanation! I will update the command.

Author: Done in e933579e5f9a007f4e406f422a5d4f0933b62e2c

>     In a package environment the ```@infiltrate``` macro only works when used in a specific way, 
>     which is why the functional form
>     ```julia
>     if isdefined(Main, :Infiltrator)
>     Main.infiltrate(@__MODULE__, Base.@locals, @__FILE__, @__LINE__)
>     end
>     ```
>     should be used to set breakpoints when working with Trixi or other packages.

Reviewer: ```suggestion
    When running Julia insdie a package environment, the ```@infiltrate``` macro only works
    if `Infiltrator` has been added to the dependencies. If this is not desired,
    the functional form
    ```julia
    if isdefined(Main, :Infiltrator)
    Main.infiltrate(@__MODULE__, Base.@locals, @__FILE__, @__LINE__)
    end
    ```
    can be used to set breakpoints when working with Trixi or other packages.
```

Author: There's also a second option with Revise, that I would mention. I am not sure though if several options might be a bit confusing for new users, which is why I did not specify them before. What do you think about that?

```suggestion
    When running in a package environment the ```@infiltrate``` macro only works if `Infiltrator` 
    has been added to the dependencies. Another work around when using Revise is to first load the
    package and then add breakpoints with `Main.@infiltrate` to the code. If this is not 
    desired, the functional form
    ```julia
    if isdefined(Main, :Infiltrator)
    Main.infiltrate(@__MODULE__, Base.@locals, @__FILE__, @__LINE__)
    end
    ```
    can be used to set breakpoints when working with Trixi or other packages.
```

Reviewer: Oh, I like yours much better, let's use your suggestion!

Author: Okay, I added my new suggestion in e933579e5f9a007f4e406f422a5d4f0933b62e2c

Triggering the breakpoint starts a REPL session where it is possible to interact with the current 
local scope. Possible commands are:
- ```@locals```: Print the local variables.
- ```@exfiltrate```: Save the local variables to a global storage, which can be accessed with 
  the ```safehouse``` variable outside the Infiltrator session.
- ```@trace```: Print the current stack trace.
- Execute other arbitrary statements
- ```?```: Print a help list with all options

> To finish the session then either use ```@continue``` to continue and eventually stop at the next 
> breakpoint or ```@exit``` to skip further breakpoints. After the code has finished local variables 
> saved with ```@exfiltrate``` can be accessed in the REPL using the ```safehouse``` variable.

Reviewer: ```suggestion
To finish a debugging session, either use ```@continue``` to continue and eventually stop at the next 
breakpoint or ```@exit``` to skip further breakpoints. After the code has finished, local variables 
saved with ```@exfiltrate``` can be accessed in the REPL using the ```safehouse``` variable.
```

Author: Done in e933579e5f9a007f4e406f422a5d4f0933b62e2c

> #### Limitations and Issues
> - Local variables cannot be changed
> - It is not possible to step into further calls or access other function scopes

Reviewer: ```suggestion
Limitations of using Infiltrator.jl are that local variables cannot be changed, and that
it is not possible to step into further calls or access other function scopes.
```

Author: Done in e933579e5f9a007f4e406f422a5d4f0933b62e2c

mode [julia-emacs](https://github.com/JuliaEditorSupport/julia-emacs).

## Debugging
Julia offers several options for debugging. A classical debugger is available with the 
[Debugger.jl](https://github.com/JuliaDebug/Debugger.jl) package or in the 
[Julia extension for VS Code](https://www.julia-vscode.org/docs/stable/userguide/debugging/). 
However, it can be quite slow and, at the time of writing (January 2023), does currently not work 

> properly with Trixi. The [Infiltrator.jl](https://github.com/JuliaDebug/Infiltrator.jl) package on 

Reviewer 3: Nevermind...sorry

> However, it can be quite slow and, at the time of writing (January 2023), does currently not work 
> properly with Trixi. The [Infiltrator.jl](https://github.com/JuliaDebug/Infiltrator.jl) package on 

Reviewer 3: ```suggestion
However, it can be quite slow and, at the time of writing (January 2023), currently does not work 
properly with Trixi. The [Infiltrator.jl](https://github.com/JuliaDebug/Infiltrator.jl) package on 
```

the other hand does not offer all features of a full debugger, but is a fast and simple tool that 
allows users to set breakpoints to open a local REPL session and access the call stack and variables.

### Infiltrator
The Infiltrator package provides fast, interactive breakpoints using the ```@infiltrate``` command, 
which drops the user into a local REPL session. From there, it is possible to access local variables, 

> see the call stack and execute statements.

Reviewer 2: ```suggestion
see the call stack, and execute statements.
```
